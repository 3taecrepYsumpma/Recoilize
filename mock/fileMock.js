// mock used for Jest config in package.json
module.exports = 'test-file-stub';