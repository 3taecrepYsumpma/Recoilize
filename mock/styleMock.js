// mock for jest CSS files, used in package.json for jest config
module.exports = {};