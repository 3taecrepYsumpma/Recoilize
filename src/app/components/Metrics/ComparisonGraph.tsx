import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';
import {dataDurationArr} from '../../../types';
import {useAppSelector} from '../../state-management/hooks';

interface ComparisonGraphProps {
  data: dataDurationArr; // an array of object{name:, actualDuration}
  width?: number;
  height?: number;
}

const ComparisonGraph: React.FC<ComparisonGraphProps> = ({
  data,
  width,
  height,
}: ComparisonGraphProps) => {
  const snapshotHistory = useAppSelector(
    (state: {snapshot: {snapshotHistory: any}}) =>
      state.snapshot.snapshotHistory,
  );
  console.log('comparison snapshot ', snapshotHistory);
  // declare an array that holds 2 objects: past and current
  const displayData = [
    {name: 'past', duration: 0},
    {name: 'current', duration: 0},
  ];

  // retrieve and get total duration for past serie from the local storage
  const values: any[] = [];
  const keys = Object.keys(localStorage);
  let i = keys.length;
  while (i--) {
    const series = localStorage.getItem(keys[i]);
    values.push(JSON.parse(series));
  }
  for (const element of values) {
    displayData[0].duration += element.componentAtomTree.treeBaseDuration;
  }

  let total = 0;
  for (const element of snapshotHistory) {
    total += element.componentAtomTree.treeBaseDuration;
  }
  displayData[1].duration = total;
  // delete series in local storage
  const deleteSeries = () => {
    for (const i of keys) {
      localStorage.removeItem(i);
    }
  };

  // svg
  const svgRef = useRef();
  useEffect(() => {
    document.getElementById('canvas').innerHTML = '';
    // set the dimensions and margins of the graph
    let left = 80;
    if (length > 13) {
      left = 100;
    }
    if (length > 17) {
      left = 120;
    }

    const margin = {top: 20, right: 20, bottom: 30, left};
    // set range for y scale
    const y = d3.scaleBand().range([height, 0]).padding(0.2);
    // set range for x scale
    const x = d3.scaleLinear().range([0, width * 0.8]);
    // set range for durations
    const z = d3.scaleBand().range([height, 0]).padding(0.2);
    // determines the color based on actualDuration
    function colorPicker(data: any) {
      if (data <= 1) return '#51a8f0';
      else if (data <= 2) return '#3a7bb0';
      else return '#2d608a';
    }
    // append the svg object to the body of the page
    // append a 'group' element to 'svg'
    // moves the 'group' element to the top left margin
    const svg = d3
      .select(svgRef.current)
      .classed('svg-container', true)
      .append('svg')
      .attr('class', 'chart')
      .attr('viewBox', '-100 0 900 600')
      .attr('preserveAspectRatio', 'xMinYMin meet')
      .classed('svg-content-responsive', true)
      .call(
        d3.zoom().on('zoom', function () {
          svg.attr('transform', d3.event.transform);
        }),
      )
      .append('g')
      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
    // Scale the range of the data in the domain
    x.domain([
      0,
      d3.max(displayData, (d: any) => {
        return d.duration;
      }),
    ]);
    // Scale the range of the data across the y-axis
    y.domain(
      displayData.map((d: any, i) => {
        return d.name + '-' + i;
      }),
    );
    // Scale actualDuration with the y-axis
    z.domain(
      displayData.map((d: any, i) => {
        return d.duration.toFixed(2) + 'ms' + '-' + i;
      }),
    );
    // append the rectangles for the bar chart
    svg
      .selectAll('.bar')
      .data(displayData)
      .enter()
      .append('rect')
      .attr('class', 'bar')
      .on('mouseover', function () {
        d3.select(this).attr('opacity', '0.85');
        const backgroundConnection = chrome.runtime.connect();
        const barName = this.data;
        const payload = {
          action: 'mouseover',
          tabId: chrome.devtools.inspectedWindow.tabId,
          payload: barName,
        };
        backgroundConnection.postMessage(payload);
      })
      // .transition()
      // .duration(1300)
      // .delay((d: any,i: any) => i * 100)
      .attr('width', function (d: any) {
        return x(d.duration);
      })
      .attr('fill', function (d: any) {
        return colorPicker(d.duration);
      })
      .attr('y', function (d: any, i: any) {
        return y(d.name + '-' + i);
      })
      .attr('height', y.bandwidth());
    // add x axis
    svg
      .append('g')
      .attr('transform', 'translate(0,' + height + ')')
      .call(d3.axisBottom(x));
    // add y axis to able to have duplicate strings
    const yAxis = d3.axisLeft(y).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    yAxis(svg.append('g'));
    // add z axis to display all duration times
    const zAxis = d3.axisRight(z).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    zAxis(svg.append('g'));
    // svg.append('g')
    //   .call(d3.axisRight(z));
  }, [data]);

  return (
    <div data-testid="canvas" id="stateGraphContainer">
      <button
        id="docs_button"
        onClick={() => {
          deleteSeries();
        }}>
        Delete Series
      </button>
      <svg id="canvas" ref={svgRef}></svg>
    </div>
  );
};

export default ComparisonGraph;
