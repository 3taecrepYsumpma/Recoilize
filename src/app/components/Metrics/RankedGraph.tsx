import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';
import {dataDurationArr} from '../../../types';

interface RankedGraphProps {
  data: dataDurationArr;
  width?: number;
  height?: number;
}

const RankedGraph: React.FC<RankedGraphProps> = ({
  data,
  width,
  height,
}: RankedGraphProps) => {
  const svgRef = useRef();
  useEffect(() => {
    document.getElementById('canvas').innerHTML = '';
    // set the dimensions and margins of the graph
    let left = 80;
    if (length > 13) {
      left = 100;
    }
    if (length > 17) {
      left = 120;
    }

    const margin = {top: 20, right: 20, bottom: 30, left};
    // set range for y scale
    const y = d3.scaleBand().range([height, 0]).padding(0.2);
    // set range for x scale
    const x = d3.scaleLinear().range([0, width * 0.8]);
    // set range for durations
    const z = d3.scaleBand().range([height, 0]).padding(0.2);
    // determines the color based on actualDuration
    function colorPicker(data: any) {
      if (data <= 1) return '#51a8f0';
      else if (data <= 2) return '#3a7bb0';
      else return '#2d608a';
    }
    // append the svg object to the body of the page
    // append a 'group' element to 'svg'
    // moves the 'group' element to the top left margin
    const svg = d3
      .select(svgRef.current)
      .classed('svg-container', true)
      .append('svg')
      .attr('class', 'chart')
      .attr('viewBox', '0 0 600 490')
      .attr('preserveAspectRatio', 'xMinYMin meet')
      .classed('svg-content-responsive', true)
      .call(
        d3.zoom().on('zoom', function () {
          svg.attr('transform', d3.event.transform);
        }),
      )
      .append('g')
      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
    // Scale the range of the data in the domain
    x.domain([
      0,
      d3.max(data, (d: any) => {
        return d.actualDuration;
      }),
    ]);
    // Scale the range of the data across the y-axis
    y.domain(
      data.map((d: any, i) => {
        return d.name + '-' + i;
      }),
    );
    // Scale actualDuration with the y-axis
    z.domain(
      data.map((d: any, i) => {
        return d.actualDuration.toFixed(2) + 'ms' + '-' + i;
      }),
    );
    // append the rectangles for the bar chart
    svg
      .selectAll('.bar')
      .data(data)
      .enter()
      .append('rect')
      .attr('class', 'bar')
      .on('mouseover', function () {
        d3.select(this).attr('opacity', '0.85');
        const backgroundConnection = chrome.runtime.connect();
        const barName = this.data;
        const payload = {
          action: 'mouseover',
          tabId: chrome.devtools.inspectedWindow.tabId,
          payload: barName,
        };
        backgroundConnection.postMessage(payload);
      })
      // .transition()
      // .duration(1300)
      // .delay((d: any,i: any) => i * 100)
      .attr('width', function (d: any) {
        return x(d.actualDuration);
      })
      .attr('fill', function (d: any) {
        return colorPicker(d.actualDuration);
      })
      .attr('y', function (d: any, i: any) {
        return y(d.name + '-' + i);
      })
      .attr('height', y.bandwidth());
    // add x axis
    svg
      .append('g')
      .attr('transform', 'translate(0,' + height + ')')
      .call(d3.axisBottom(x));
    // add y axis to able to have duplicate strings
    const yAxis = d3.axisLeft(y).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    yAxis(svg.append('g'));
    // add z axis to display all duration times
    const zAxis = d3.axisRight(z).tickFormat(function (d: any) {
      return d.split('-')[0];
    });
    zAxis(svg.append('g'));
    // svg.append('g')
    //   .call(d3.axisRight(z));
  }, [data]);

  return (
    <div data-testid="canvas" id="stateGraphContainer">
      <button
        className="save-series-button"
        onClick={e => {
          console.log('save button has been clicked');
        }}>
        Save Series
      </button>
      <svg id="canvas" ref={svgRef}></svg>
    </div>
  );
};

export default RankedGraph;
