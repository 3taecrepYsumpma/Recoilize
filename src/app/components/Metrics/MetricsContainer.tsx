import React, {useState} from 'react';
import {ParentSize} from '@vx/responsive';
import {dataDurationArr} from '../../../types';
import FlameGraph from './FlameGraph.js';
import RankedGraph from './RankedGraph';
import ComparisonGraph from './ComparisonGraph';
import {useAppSelector} from '../../state-management/hooks';

const Metrics: React.FC = () => {
  const cleanedComponentAtomTree = useAppSelector(
    state => state.snapshot.cleanComponentAtomTree,
  );
  //create state for the graph type toggle
  const [graphType, setGraphType] = useState<string>('flame');

  //funciton that toggles the graphType state
  const toggleGraphFunc = (check: string): void => {
    if (check === 'flame') setGraphType('flame');
    if (check === 'ranked') setGraphType('ranked');
    if (check === 'comparison') setGraphType('comparison');
  };

  // create an empty array to store objects for property name and actualDuration for rankedGraph
  const dataDurationArr: dataDurationArr = [];
  let length = 0;
  // function to traverse through the fiber tree
  const namesAndDurations = (node: any) => {
    if (node === undefined) return;
    if (node.name && node.actualDuration) {
      const obj: any = {};
      if (node.name.length > length) {
        length = node.name.length;
      }
      obj['name'] = node.name;
      obj['actualDuration'] = node.actualDuration;
      dataDurationArr.push(obj);
    }
    node.children.forEach((child: any) => namesAndDurations(child));
  };
  namesAndDurations(cleanedComponentAtomTree);

  //function that renders either graphComponent based on graphType state variable
  const determineRender: any = () => {
    if (graphType === 'flame') {
      return (
        //ParentSize component allows us to scale the FlameGraph to fit its container.
        <ParentSize>
          {size =>
            size.ref && (
              <FlameGraph
                cleanedComponentAtomTree={cleanedComponentAtomTree}
                width={size.width}
                height={600}
              />
            )
          }
        </ParentSize>
      );
    } else if (graphType === 'ranked') {
      return (
        <ParentSize>
          {size =>
            size.ref && (
              <RankedGraph
                data={dataDurationArr}
                width={size.width}
                height={size.height}
              />
            )
          }
        </ParentSize>
      );
    } else if (graphType === 'comparison') {
      return (
        <ParentSize>
          {size =>
            size.ref && (
              <ComparisonGraph
                data={dataDurationArr}
                width={size.width}
                height={size.height}
              />
            )
          }
        </ParentSize>
      );
    }
  };

  //render the toggle buttons and the appropriate graph based on GraphType state variable
  return (
    <div id="metricsWrapper">
      <div data-testid="canvas" className="graphContainer">
        <button
          className="graphButton"
          autoFocus={true}
          onClick={() => {
            toggleGraphFunc('flame');
          }}
        >
          Flame Graph
        </button>
        <button
          className="graphButton"
          onClick={() => {
            toggleGraphFunc('ranked');
          }}
        >
          Ranked Graph
        </button>
        <button
          className="graphButton"
          onClick={() => {
            toggleGraphFunc('comparison');
          }}
        >
          Comparison Graph
        </button>
      </div>
      {determineRender()}
    </div>
  );
};

export default Metrics;
