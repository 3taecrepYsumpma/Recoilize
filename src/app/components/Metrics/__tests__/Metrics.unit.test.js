import React from 'react';
import '@testing-library/jest-dom'

import {render, generateStore, screen, cleanup} from '../../../tests/testing';

import Metrics from '../MetricsContainer';

// this test is loosely based on this article on rendering React tests https://betterprogramming.pub/react-testing-library-configuration-for-productive-unit-testing-5d0c446f3b3d

// this is our mock state that we will use to run our tests
import { snapshotHistoryMock } from '../../../../../mock/state-snapshot';
// import React from 'react';
// import Metrics from '../MetricsContainer';
// import {render, cleanup} from '@testing-library/react';
// import '@testing-library/dom';
// import '@testing-library/jest-dom'

// import {screen} from '../../../tests/testing';
// import { snapshotHistoryMock } from '../../../../../mock/state-snapshot';

afterEach(cleanup);
const mock = snapshotHistoryMock.cleanComponentAtomTree;
// TO-DO
// import prop object being passed down to metric graphs to be tested
describe('Metrics rendering', () => {
  it('should render properly', () => {
    const store = generateStore({ snapshot: snapshotHistoryMock})
    const asFragment = render(
        <Metrics/>, {providers: {store}})
      expect(asFragment).toMatchSnapshot();
  });
  xit('should have cleanComponentAtomTree, data, height, and width', () => {
    const store = generateStore({ snapshot: snapshotHistoryMock})
    const asFragment = render(
        <Metrics/>, {providers: {store}})
      expect(asFragment).toMatchSnapshot();
  });
});

// })
// //create a shallow copy of Metrics component passing in expected prop
// const wrapper = shallow(<Metrics includedProp={cleanedComponentAtomTree}/>);
// describe('shape of props being passed down should match cleanedComponentAtomTree shape', () => {
//     expect(wrapper.props().includedProp).toMatchObject(componentAtomTree);
// });
// describe('component is being rendered correctly', () => {
//     const { getByTestId } = render(<Metrics />);
//     //if the component rendered then it has to have return an element with id 'canvas'
//     //getBy returns an error when not finding an element, here is is looking for an id of canvas
//     expect(getByTestId('canvas')).toBeTruthy();