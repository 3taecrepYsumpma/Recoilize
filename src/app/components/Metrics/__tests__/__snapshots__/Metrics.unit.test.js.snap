// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics rendering should render properly 1`] = `undefined`;
